panther_analysis_tool/testing.py

```diff
@@ -0,0 +1,263 @@
+    name: str
+    data: Dict[str, Any]
+    mocks: List[Dict[str, Any]]
+    expectations: TestExpectations
+
+
+# pylint: disable=too-few-public-methods
+class TestCaseEvaluator:
+    """Translates detection execution results to test case results,
+    by performing assertions and determining the status"""
+
+    def __init__(self, spec: TestSpecification, detection_result: DetectionResult):
+        self._spec = spec
+        self._detection_result = detection_result
+
+    def _get_result_status(self) -> bool:
+        """Get the test status - passing/failing"""
+
+        # matched attribute can also be None,
+        # coerce to boolean for consistent return values
+        matched = (
+            bool(self._detection_result.matched)
+            == self._detection_result.detection_match_alert_value
+        )
+
+        # Title/dedup functions are executed unconditionally
+        # (regardless if the detection matched or not) during testing.
+        # Only if the detection is expected to trigger an alert,
+        # we want to include errors from other functions in the status.
+        if self._spec.expectations.detection == self._detection_result.detection_match_alert_value:
+            # Any error should mark the test as failing
+            return matched and not self._detection_result.errored
+
+        # Only detection/setup exceptions and event compatibility (JSON-decodable and JSON object)
+        # should be a factor in marking the test as failing
+        return (
+            self._detection_result.input_exception is None
+            and not matched
+            and not self._detection_result.detection_evaluation_failed
+        )
+
+    def _get_generic_error_details(self) -> Tuple[Optional[Exception], Optional[str]]:
+        generic_error = None
+        generic_error_title = None
+        if self._detection_result.input_exception is not None:
+            generic_error = self._detection_result.input_exception
+            generic_error_title = "Invalid event"
+        elif self._detection_result.setup_exception is not None:
+            generic_error = self._detection_result.setup_exception
+        return generic_error, generic_error_title
+
+    def interpret(self) -> TestResult:
+        """Evaluate the detection result taking into account
+        the errors raised during evaluation and
+        the test specification expectations"""
+
+        function_results = dict(
+            detectionFunction=FunctionTestResult.new(
+                self._detection_result.matched, self._detection_result.detection_exception
+            )
+        )
+
+        # We don't include output from other functions
+        # unless the test was expected to match and trigger an alert.
+        # Even if the test fails, providing all the output provides a faster feedback loop,
+        # on possible additional failures.
+        if self._spec.expectations.detection == self._detection_result.detection_match_alert_value:
```

Comment:
Maybe use the `is` operator here since we expect both values to be strictly boolean? It will be easier to reason about.